File: Gio/src/main/java/com/hd/plugin/lifecycle/LifecycleClassVisitor.java
package com.hd.plugin.lifecycle;

import groovyjarjarasm.asm.ClassVisitor;
import groovyjarjarasm.asm.MethodVisitor;
import groovyjarjarasm.asm.Opcodes;

public class LifecycleClassVisitor extends ClassVisitor implements Opcodes {
    private String mClassName;

    public LifecycleClassVisitor(ClassVisitor classVisitor) {
        super(Opcodes.ASM5, classVisitor);
    }

    @Override
    public void visit(int version, int access, String name, String signature, String superName, String[] interfaces) {
        System.out.println("LifecycleClassVisitor : visit -----> started :" + name);
        mClassName = name;
        super.visit(version, access, name, signature, superName, interfaces);
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
        System.out.println("LifecycleClassVisitor : visitMethod : " + name);
        MethodVisitor methodVisitor = cv.visitMethod(access, name, descriptor, signature, exceptions);
        if (mClassName.equals("androidx/fragment/app/FragmentActivity")) {
            if (name.equals("onCreate")) {
                System.out.println("LifecycleClassVisitor : change method ----> " + name);
                return new LifecycleOnCreateMethodVisitor(methodVisitor);
            }
        }
        return methodVisitor;
    }

    @Override
    public void visitEnd() {
        System.out.println("Gio : visit -----> end");
        super.visitEnd();
    }
}

File: buildSrc/src/main/java/com/hd/plugin/lifecycle/LifecycleClassVisitor.kt
package com.hd.plugin.lifecycle

import groovyjarjarasm.asm.ClassVisitor
import groovyjarjarasm.asm.MethodVisitor
import groovyjarjarasm.asm.Opcodes

class LifecycleClassVisitor(classVisitor: ClassVisitor?) :
    ClassVisitor(Opcodes.ASM5, classVisitor),
    Opcodes {
    private var mClassName: String? = null
    override fun visit(version: Int, access: Int, name: String, signature: String, superName: String, interfaces: Array<String>) {
        println("LifecycleClassVisitor : visit -----> started :$name")
        mClassName = name
        super.visit(version, access, name, signature, superName, interfaces)
    }

    override fun visitMethod(access: Int, name: String, descriptor: String, signature: String, exceptions: Array<String>): MethodVisitor {
        println("LifecycleClassVisitor : visitMethod : $name")
        val methodVisitor =
            cv.visitMethod(access, name, descriptor, signature, exceptions)
        if (mClassName == "androidx/fragment/app/FragmentActivity") {
            if (name == "onCreate") {
                println("LifecycleClassVisitor : change method ----> $name")
                return LifecycleOnCreateMethodVisitor(methodVisitor)
            }
        }
        return methodVisitor
    }

    override fun visitEnd() {
        println("Gio : visit -----> end")
        super.visitEnd()
    }
}